File: src/sql_parser.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../include/sql_parsed_command.h"
#include "../include/sql_parser.h"
#include "../include/statements.h"
#include "../include/str_split.h"
#include "../include/list.h"

static SQLParsedCommand* parse_insert(SQLParsedCommand *parsed_command, List *tokens);
static SQLParsedCommand* parse_select(List *tokens);
static SQLParsedCommand* parse_delete(List *tokens);

SQLParsedCommand* sql_parser(char *buffer, Statement *statement) {
    SQLParsedCommand *parsed_command = new_sql_parsed_command();
    
    List *tokens = str_split(buffer, " ");
    
    switch (statement->type) {
        case STATEMENT_INSERT:
            parsed_command = parse_insert(parsed_command, tokens);
            break;
        case STATEMENT_SELECT:
            parsed_command = parse_select(tokens);
            break;
        case STATEMENT_DELETE:
            parsed_command = parse_delete(tokens);
            break;
    }

    free_list(tokens);

    return parsed_command;
}

static SQLParsedCommand* parse_insert(SQLParsedCommand *parsed_command, List *tokens) {
    if (tokens->size < 4 || strcmp(get_element_at(tokens, 1), "INTO") != 0
    || strcmp(get_element_at(tokens, 3), "VALUES") != 0) {
        printf("Invalid INSERT command syntax!\n");
        return NULL;
    }

    push_back(parsed_command->tables, 
              strdup(get_element_at(tokens, 2)));

    for (size_t i = 4; i < tokens->size; i++) {
        push_back(parsed_command->values, 
                  strdup(get_element_at(tokens, i)));
    }

    return parsed_command;
}

static SQLParsedCommand* parse_delete(List *tokens) {
}

static SQLParsedCommand* parse_select(List *tokens) {
}

File: internal/lib/sqlparser/sqlparser.go
package sqlparser

import (
	"errors"
	"strings"
)

type SQLCommandType int

const (
	InsertCommand SQLCommandType = iota
	SelectCommand
	DeleteCommand
)

type SQLParsedCommand struct {
	Type      SQLCommandType
	Tables    []string
	Columns   []string
	Values    []string
	Condition string
}

func ParseSQL(input string) (*SQLParsedCommand, error) {
	tokens := strings.Fields(input)
	if len(tokens) == 0 {
		return nil, errors.New("empty input")
	}

	commandType, err := detectCommandType(tokens[0])
	if err != nil {
		return nil, err
	}

	parsedCommand := &SQLParsedCommand{Type: commandType}

	switch commandType {
	case InsertCommand:
		return parseInsert(tokens, parsedCommand)
	case SelectCommand:
		return parseSelect(tokens, parsedCommand)
	case DeleteCommand:
		return parseDelete(tokens, parsedCommand)
	}

	return nil, errors.New("unsupported command type")
}

func detectCommandType(command string) (SQLCommandType, error) {
	switch strings.ToUpper(command) {
	case "INSERT":
		return InsertCommand, nil
	case "SELECT":
		return SelectCommand, nil
	case "DELETE":
		return DeleteCommand, nil
	default:
		return -1, errors.New("unrecognized command type")
	}
}

func parseInsert(tokens []string, parsedCommand *SQLParsedCommand) (*SQLParsedCommand, error) {
	if len(tokens) < 4 || strings.ToUpper(tokens[1]) != "INTO" || strings.ToUpper(tokens[3]) != "VALUES" {
		return nil, errors.New("invalid INSERT command syntax")
	}

	parsedCommand.Tables = append(parsedCommand.Tables, tokens[2])
	values := strings.Trim(tokens[4], "()")
	parsedCommand.Values = splitAndTrim(values, ",")

	return parsedCommand, nil
}

func parseSelect(tokens []string, parsedCommand *SQLParsedCommand) (*SQLParsedCommand, error) {
	fromIndex := findKeywordIndex(tokens, "FROM")
	if fromIndex == -1 || fromIndex == 1 {
		return nil, errors.New("invalid SELECT command syntax")
	}

	parsedCommand.Columns = splitAndTrim(strings.Join(tokens[1:fromIndex], " "), ",")
	whereIndex := findKeywordIndex(tokens, "WHERE")

	if whereIndex == -1 {
		parsedCommand.Tables = tokens[fromIndex+1:]
	} else {
		parsedCommand.Tables = tokens[fromIndex+1 : whereIndex]
		parsedCommand.Condition = strings.Join(tokens[whereIndex+1:], " ")
	}

	return parsedCommand, nil
}

func parseDelete(tokens []string, parsedCommand *SQLParsedCommand) (*SQLParsedCommand, error) {
	if len(tokens) < 3 || strings.ToUpper(tokens[1]) != "FROM" {
		return nil, errors.New("invalid DELETE command syntax")
	}

	parsedCommand.Tables = append(parsedCommand.Tables, tokens[2])
	if len(tokens) > 3 && strings.ToUpper(tokens[3]) == "WHERE" {
		parsedCommand.Condition = strings.Join(tokens[4:], " ")
	}

	return parsedCommand, nil
}

func findKeywordIndex(tokens []string, keyword string) int {
	for i, token := range tokens {
		if strings.ToUpper(token) == keyword {
			return i
		}
	}
	return -1
}

func splitAndTrim(input, delimiter string) []string {
	parts := strings.Split(input, delimiter)
	for i, part := range parts {
		parts[i] = strings.TrimSpace(part)
	}
	return parts
}
